File: src/main/java/com/godwin/jsonparser/util/JsonUtils.java
package com.godwin.jsonparser.util;

import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.util.DefaultIndenter;
import com.fasterxml.jackson.core.util.DefaultPrettyPrinter;
import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.List;
import java.util.Map;

public class JsonUtils {

    private JsonUtils() {
    }

    public static String formatJson(String jsonStr) throws JsonProcessingException {
        ObjectMapper mapper = Holder.MAPPER;
        mapper.configure(JsonParser.Feature.ALLOW_UNQUOTED_FIELD_NAMES, true);
        mapper.configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true);
        mapper.configure(JsonParser.Feature.ALLOW_COMMENTS, true);
        mapper.configure(JsonParser.Feature.ALLOW_YAML_COMMENTS, true);
        Object jsonObject = mapper.readValue(jsonStr, Object.class);
        return Holder.MAPPER.writer(Holder.DEFAULT_PRETTY_PRINTER).writeValueAsString(jsonObject);
    }

    public static Object getMap(String jsonStr) throws JsonProcessingException {
        ObjectMapper mapper = Holder.MAPPER;
        mapper.configure(JsonParser.Feature.ALLOW_UNQUOTED_FIELD_NAMES, true);
        mapper.configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true);
        mapper.configure(JsonParser.Feature.ALLOW_COMMENTS, true);
        mapper.configure(JsonParser.Feature.ALLOW_YAML_COMMENTS, true);

        JsonNode node = mapper.readTree(jsonStr);

        if (node.isObject()) {
            // If it's a JSON object, map to a Map
            return mapper.convertValue(node, Map.class);
        } else if (node.isArray()) {
            // If it's a JSON array, map to a List of Maps
            return mapper.convertValue(node, List.class);
        } else {
            // Handle unexpected cases (e.g., primitive values)
            throw new IllegalArgumentException("Unsupported JSON structure");
        }
    }

    public static Object parseJson(String json, ObjectMapper objectMapper) throws Exception {
        // Parse the JSON into a JsonNode
        JsonNode node = objectMapper.readTree(json);

        if (node.isObject()) {
            // If it's a JSON object, map to a Map
            return objectMapper.convertValue(node, Map.class);
        } else if (node.isArray()) {
            // If it's a JSON array, map to a List of Maps
            return objectMapper.convertValue(node, List.class);
        } else {
            // Handle unexpected cases (e.g., primitive values)
            throw new IllegalArgumentException("Unsupported JSON structure");
        }
    }

    /**
     * Cleans up the JSON string by applying various cleanup methods.
     * For now, it only ensures missing opening and closing braces are fixed.
     * Other future methods that can be added are:
     * 1. Missing only opening brace
     * 2. Missing only closing brace
     * 2. Missing comma
     * 3. missing '"'
     */
    public static String cleanUpJsonString(String jsonString) {
        jsonString = jsonString.trim();

        // If it's already valid, return as is.
        if (isValidJson(jsonString)) {
            return jsonString;
        }

        //attempt to fix json

        // Check if the start or end braces is missing
        if (!jsonString.startsWith("{") || !jsonString.endsWith("}")) {
            String fixedJsonString = fixMissingBraces(jsonString);

            // Validate after adding missing braces
            if (isValidJson(fixedJsonString)) {
                return fixedJsonString;
            }
        }

        //Return the JSON in the original form
        return jsonString;
    }

    /**
     * Checks if the given JSON string is valid.
     * @param jsonString The JSON string to validate.
     * @return true if the JSON is valid, false otherwise.
     */
    private static boolean isValidJson(String jsonString) {
        try (JsonParser parser = new JsonFactory().createParser(jsonString)) {
            //noinspection StatementWithEmptyBody
            while (parser.nextToken() != null) {
                // If we can iterate through tokens then it's valid JSON
            }
            return true;
        } catch (Exception e) {
            return false; // Invalid JSON
        }
    }

    /**
     * Fixes missing opening and closing braces in the JSON string.
     * @param jsonString The JSON string to fix.
     * @return A properly wrapped JSON string.
     */
    private static String fixMissingBraces(String jsonString) {
        jsonString = jsonString.trim();

        int openBracesCount = 0;
        int closeBracesCount = 0;

        // Count `{` and `}` braces
        for (char ch : jsonString.toCharArray()) {
            if (ch == '{') openBracesCount++;
            if (ch == '}') closeBracesCount++;
        }

        if (openBracesCount == closeBracesCount && !jsonString.startsWith("{")) {
            jsonString = "{" + jsonString + "}";
        }

        return jsonString;
    }

/*    public static String minifyJson(String jsonStr) throws JsonProcessingException {
        Object jsonObject = Holder.MAPPER.readValue(jsonStr, Object.class);
        return Holder.MAPPER.writeValueAsString(jsonObject);
    }*/

/*    public static void verifyJson(String jsonStr) throws JsonProcessingException {
        Holder.MAPPER.readValue(jsonStr, Object.class);
    }*/

    private static final class Holder {
        public static final ObjectMapper MAPPER = new CustomMapper();
        public static final DefaultPrettyPrinter DEFAULT_PRETTY_PRINTER = new CustomPrettyPrinter();
    }

    private static final class CustomPrettyPrinter extends DefaultPrettyPrinter {

        private static final DefaultIndenter UNIX_LINE_FEED_INSTANCE = new DefaultIndenter("  ", "\n");

        public CustomPrettyPrinter() {
            super._objectFieldValueSeparatorWithSpaces = ":";
            super._objectIndenter = UNIX_LINE_FEED_INSTANCE;
            super._arrayIndenter = UNIX_LINE_FEED_INSTANCE;
            super._spacesInObjectEntries = true;
        }

        @Override
        public DefaultPrettyPrinter createInstance() {
            return new CustomPrettyPrinter();
        }
    }

    public static final class CustomMapper extends ObjectMapper {
        public CustomMapper() {
            configure(DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES, false);
            configure(DeserializationFeature.FAIL_ON_NUMBERS_FOR_ENUMS, false);
            configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
        }
    }
}

File: src/main/kotlin/com/godwin/jsonparser/util/JsonUtils.kt
package com.godwin.jsonparser.util

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.core.util.DefaultIndenter
import com.fasterxml.jackson.core.util.DefaultPrettyPrinter
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.SerializationFeature
import com.fasterxml.jackson.dataformat.yaml.YAMLMapper
import com.godwin.jsonparser.services.JsonPersistence

object JsonUtils {

    private val mapper = CustomMapper().apply {
        configure(JsonParser.Feature.ALLOW_UNQUOTED_FIELD_NAMES, true)
        configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true)
        configure(JsonParser.Feature.ALLOW_COMMENTS, true)
        configure(JsonParser.Feature.ALLOW_YAML_COMMENTS, true)
    }

    fun formatJson(jsonStr: String): String {
        val prefs = JsonPersistence.getInstance()
        val printer = CustomPrettyPrinter(prefs.indentSize)
        val node = mapper.readTree(jsonStr)
        val writeMapper = if (prefs.sortKeys) {
            mapper.copy().configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true)
        } else mapper
        return writeMapper.writer(printer).writeValueAsString(node)
    }

    fun isValidJson(jsonStr: String): Boolean {
        return try {
            mapper.readTree(jsonStr)
            true
        } catch (_: Exception) {
            false
        }
    }

    fun isValidJsonError(jsonStr: String): String? {
        return try {
            ObjectMapper().readTree(jsonStr)
            null
        } catch (e: Exception) {
            return if (e is JsonProcessingException) {
                e.originalMessage
            } else {
                e.message
            }
        }
    }


    fun getMap(jsonStr: String): Any {
        val node = mapper.readTree(jsonStr)

        return when {
            node.isObject -> mapper.convertValue(node, Map::class.java)
            node.isArray -> mapper.convertValue(node, List::class.java)
            else -> throw IllegalArgumentException("Unsupported JSON structure")
        }
    }

    fun minifyJson(jsonStr: String): String {
        val node = mapper.readTree(jsonStr)
        return mapper.writeValueAsString(node)
    }

    fun queryJsonPath(jsonStr: String, expression: String): String {
        val result = com.jayway.jsonpath.JsonPath.read<Any>(jsonStr, expression)
        return mapper.writer(CustomPrettyPrinter()).writeValueAsString(result)
    }

    fun queryJmesPath(jsonStr: String, expression: String): String {
        val runtime = io.burt.jmespath.jackson.JacksonRuntime()
        val expr = runtime.compile(expression)
        val node = mapper.readTree(jsonStr)
        val result = expr.search(node)
        return mapper.writer(CustomPrettyPrinter()).writeValueAsString(result)
    }

    fun toYaml(jsonStr: String): String {
        val node = mapper.readTree(jsonStr)
        return YAMLMapper().writeValueAsString(node)
    }

    fun toJsonSchema(jsonStr: String): String {
        val node = mapper.readTree(jsonStr)
        val schema = inferSchema(node, "root")
        return mapper.writer(CustomPrettyPrinter()).writeValueAsString(schema)
    }

    data class JsonStats(
        val keys: Int,
        val depth: Int,
        val objects: Int,
        val arrays: Int,
        val nulls: Int,
        val strings: Int,
        val numbers: Int,
        val booleans: Int,
        val sizeBytes: Int
    )

    fun computeStats(jsonStr: String): JsonStats {
        val node = mapper.readTree(jsonStr)
        var keys = 0; var maxDepth = 0; var objects = 0
        var arrays = 0; var nulls = 0; var strings = 0
        var numbers = 0; var booleans = 0

        fun walk(n: JsonNode, depth: Int) {
            if (depth > maxDepth) maxDepth = depth
            when {
                n.isObject -> { objects++; n.fields().forEach { (_, v) -> keys++; walk(v, depth + 1) } }
                n.isArray -> { arrays++; n.forEach { walk(it, depth + 1) } }
                n.isNull -> nulls++
                n.isTextual -> strings++
                n.isNumber -> numbers++
                n.isBoolean -> booleans++
            }
        }
        walk(node, 0)
        return JsonStats(keys, maxDepth, objects, arrays, nulls, strings, numbers, booleans, jsonStr.toByteArray().size)
    }

    private fun inferSchema(node: JsonNode, title: String): Map<String, Any> {
        return when {
            node.isObject -> {
                val props = mutableMapOf<String, Any>()
                node.fields().forEach { (k, v) -> props[k] = inferSchema(v, k) }
                mapOf("type" to "object", "title" to title, "properties" to props)
            }
            node.isArray -> {
                val items = if (node.size() > 0) inferSchema(node[0], "item") else mapOf("type" to "object")
                mapOf("type" to "array", "items" to items)
            }
            node.isTextual -> mapOf("type" to "string")
            node.isInt || node.isLong -> mapOf("type" to "integer")
            node.isNumber -> mapOf("type" to "number")
            node.isBoolean -> mapOf("type" to "boolean")
            node.isNull -> mapOf("type" to "null")
            else -> mapOf("type" to "string")
        }
    }

    fun cleanUpJsonString(jsonString: String): String {
        val trimmed = jsonString.trim()

        if (isValidJson(trimmed)) {
            return trimmed
        }

        if (!trimmed.startsWith("{") || !trimmed.endsWith("}")) {
            val fixed = fixMissingBraces(trimmed)
            if (isValidJson(fixed)) {
                return fixed
            }
        }

        return trimmed
    }

    private fun fixMissingBraces(jsonString: String): String {
        var result = jsonString.trim()
        var openBracesCount = 0
        var closeBracesCount = 0
        var inString = false

        result.forEachIndexed { i, ch ->
            if (ch == '"') {
                val isEscaped = i > 0 && result[i - 1] == '\\'
                if (!isEscaped) {
                    inString = !inString
                }
            }

            if (!inString) {
                when (ch) {
                    '{' -> openBracesCount++
                    '}' -> closeBracesCount++
                }
            }
        }

        if (openBracesCount == closeBracesCount && !result.startsWith("{")) {
            result = "{$result}"
        }

        return result
    }

    private class CustomPrettyPrinter(indentSize: Int = 2) : DefaultPrettyPrinter() {
        init {
            _objectFieldValueSeparatorWithSpaces = ":"
            val indent = " ".repeat(indentSize)
            _objectIndenter = DefaultIndenter(indent, "\n")
            _arrayIndenter = DefaultIndenter(indent, "\n")
        }

        override fun createInstance() = CustomPrettyPrinter()
    }

    class CustomMapper : ObjectMapper() {
        init {
            configure(DeserializationFeature.FAIL_ON_NULL_FOR_PRIMITIVES, false)
            configure(DeserializationFeature.FAIL_ON_NUMBERS_FOR_ENUMS, false)
            configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
        }
    }
}
